cibuildwheel/__main__.py

```diff
@@ -250,105 +202,20 @@ def main() -> None:
     )
     test_selector = TestSelector(skip_config=test_skip)
 
-    try:
-        environment = parse_environment(environment_config)
-    except (EnvironmentParseError, ValueError):
-        print(f'cibuildwheel: Malformed environment option "{environment_config}"', file=sys.stderr)
-        traceback.print_exc(None, sys.stderr)
-        sys.exit(2)
-
-    if dependency_versions == "pinned":
-        dependency_constraints: Optional[
-            DependencyConstraints
-        ] = DependencyConstraints.with_defaults()
-    elif dependency_versions == "latest":
-        dependency_constraints = None
-    else:
-        dependency_versions_path = Path(dependency_versions)
-        dependency_constraints = DependencyConstraints(dependency_versions_path)
-
-    if test_extras:
-        test_extras = f"[{test_extras}]"
-
-    try:
-        build_verbosity = min(3, max(-3, int(build_verbosity_str)))
-    except ValueError:
-        build_verbosity = 0
-
-    # Add CIBUILDWHEEL environment variable
-    # This needs to be passed on to the docker container in linux.py
-    os.environ["CIBUILDWHEEL"] = "1"
-
-    archs = Architecture.parse_config(archs_config_str, platform=platform)
+    build_options = compute_options(
+        options, args.archs, build_selector, test_selector, platform, package_dir, output_dir
+    )
```

Comment:
Hmm, this signature is a bit long. And now we have options parsing in two places, some in `__main__`, and some in `compute_options`. If we're moving some options computation out of `__main__`, we'd be best off moving it all, I think? So this function `compute_options` would take no arguments.

Then the use of these values in `__main__` would be e.g. `build_options.platform`. Which seems fine to me.

Reply:
The function compute_options needs several things from `args`. I chose this split to compute the parts of args that were unrelated to the build_options: the option file, the paths, the selectors, and args-specifics parts (platform and archs). These also happen to be the things that are not specific to a build selector - they will not be able to be overridden, for example. I can revisit it to see if any more could be moved in, but I don't think we should be passing in `args` - command line argument parsing should be done in `__main__`, this function is about reading the environment variables and the (previously opened) options file.

(I've already written `compute_all_options`, which runs this once per override.)

Reply:
Also, I'm intentionally avoiding multi-platform code (importing linux, macOS, windows) outside of __main__. Though I think I might be able to do a bit more, some parts I didn't want to move into a function (identifiers) aren't actually used in this function, they just happen to happen in the middle of the current code block. I'll take a look.

Reply:
Pulled things out further into two functions, take another look.

Reply:
To be honest, these options refactors still aren't really making sense to me... I think it's because they're actually motivated by the change to per-build options. But I don't want to hold you up with your work on TOML overrides. Would you be able to remove the options parsing changes from this PR and we can discuss them in the TOML override PR? I think they will make more sense in that context, so it will be easier to review.

Reply:
All the option parsing is happening in options.py now. The args parsing is the only thing left in `__main__`, because it is main's job to parse `args`, and it needs several things from args (like selecting the platform, and the package_dir).

I think this is a clear step forward without the overrides; now the `__main__.py` is shorter and imports much less stuff, and options handling is in options.py. If you wanted to write a new interface to cibuildwheel with different command line arguments, you could do it and still reuse the rest of the infrastructure. There's much less coupling between command line parsing and reading the options.

Reply:
(The overrides part is really just the breaking of `compute_options` into a sub function `_compute_single_options`, but that's internal, and still improves the grouping and readability a bit by localizing the variables.)

Reply:
(If you still want it out, I can see if I can reorder the commits and pull it out, looks doable, maybe)

Reply:
PS: I also recommend looking at the branch, rather than the diff, to see how the new `__main__.py` and `options.py` flow. I think they read better with better grouping, which is not as obvious from the diff.

Reply:
Press the "." key to drop into GitHub.dev if you want. ;)

Reply:
Splitting the args parsing out from compute_options is the thing that's giving me pause, because some of the defaults for the arguments are themselves environment variables. So it's not clear when we call compute_options, if those function arguments are supposed to act like command line arguments (and therefore override config/env vars) or if they already have those built in. But otherwise, yeah, I agree, `__main__` is looking much neater this way. 

But yeah, like I say, I don't want to hold you up, so if it's easier, I'm happy to go with this for now, and we can revisit it later if we need to.

Reply:
Okay, We can put this in, then change it as needed, maybe after finishing the overrides. I have a first version ready, but it needs quite a bit of refactoring to simplify it, and lots of tests (it passes mypy and the original unit tests so far). I'd like to merge this before making that PR to have CI run on it.